sdk/eventhub/event-hubs/src/withAuth.ts

```diff
@@ -0,0 +1,174 @@
+
+/**
+ *
+ * @param cb - The callback to be executed after the token is obtained.
+ * @param context - The connection context.
+ * @param audience - The audience for which the token is needed.
+ * @param timeoutInMs - The timeout in milliseconds.
+ * @param logger - The logger to be used for logging.
+ * @returns  A TimerLoop that keeps the token refreshed.
+ * @internal
+ */
+export async function withAuth(
+  cb: () => Promise<void>,
+  context: ConnectionContext,
+  audience: string,
+  timeoutInMs: number,
+  logger: SimpleLogger,
+  options: {
+    abortSignal?: AbortSignalLike;
+  }
+): Promise<TimerLoop> {
+  const info = await getTokenInfo(context.tokenCredential, audience);
+  await setupClaimNegotiation(context, audience, info, timeoutInMs, logger, options);
+  await cb();
+  const loop = createTimerLoop(info.timeoutInMs, () =>
+    setupClaimNegotiation(context, audience, info, timeoutInMs, logger, options)
+      .catch((err) => {
+        logger.verbose("an error occurred while renewing the token: %O", err);
+      })
+      .then(() =>
+        logger.verbose(
+          "next token renewal is in %d milliseconds @(%s).",
+          info.timeoutInMs,
+          new Date(Date.now() + info.timeoutInMs).toString()
+        )
+      )
+  );
+  loop.start();
+  return loop;
+}
+
+/**
+ * @internal
+ */
+export async function openCbsSession(
+  client: CbsClient,
+  timeoutAfterStartTime: number,
+  { abortSignal }: { abortSignal?: AbortSignalLike } = {}
+): Promise<void> {
+  return defaultCancellableLock.acquire(
+    client.cbsLock,
+    () =>
+      client.init({
+        abortSignal,
+        timeoutInMs: timeoutAfterStartTime - Date.now(),
+      }),
+    {
+      abortSignal,
+      timeoutInMs: timeoutAfterStartTime - Date.now(),
+    }
+  );
+}
+
+/**
+ * A representation of an access token.
+ * @internal
+ */
+interface TokenInfo {
+  /** The access token */
+  token: AccessToken;
+  /** The type of the token */
+  type: TokenType;
+  /** The time duration after which the token should be refreshed */
+  timeoutInMs: number;
+}
+
+async function getAadToken(cred: TokenCredential): Promise<TokenInfo> {
+  const token = await cred.getToken(Constants.aadEventHubsScope);
+  if (!token) {
+    throw new Error(`Failed to get token from the provided "TokenCredential" object`);
+  }
+  return {
+    token,
+    type: TokenType.CbsTokenTypeJwt,
+    timeoutInMs: token.expiresOnTimestamp - Date.now() - 2 * 60 * 1000,
+  };
+}
+
+function getSharedKeyBasedToken(cred: SasTokenProvider, audience: string): TokenInfo {
+  return {
+    token: cred.getToken(audience),
+    type: TokenType.CbsTokenTypeSas,
+    timeoutInMs: 45 * 60 * 1000,
+  };
+}
+
+async function getTokenInfo(
+  cred: SasTokenProvider | TokenCredential,
+  audience: string
+): Promise<TokenInfo> {
+  return isSasTokenProvider(cred)
+    ? getSharedKeyBasedToken(cred, audience)
+    : await getAadToken(cred); // eslint-disable-line no-return-await
```

Comment:
why do we need the `await` here?

Reply:
I am not familiar with how the underlying API works but it returns a promise.

Reply:
Oh I see, it went over my head that async functions wrap non-promise values in promises.

Reply:
Addressed in https://github.com/Azure/azure-sdk-for-js/pull/25214/commits/8dca899493f6d25ac2b125bbe83171f5a0a9db00.